api/README.md

```diff
@@ -0,0 +1,264 @@
+>
+> 1. 12 months or 3 releases (whichever is longer) if the API member is part of a Stable/vN version.
+> 2. 9 months or 3 releases (whichever is longer) if the API member is part of a Beta/vNbeta1 version.
+> 3. 0 releases if the API member is part of an Alpha/vNalpha1 version.
+> To continue running CSI Node Plugins that depend on an old version of csi-proxy.exe, vN, some of whose members have been removed, Kubernetes administrators will be required to run the latest version of the csi-proxy.exe (that will be used by CSI Node Plugins that use versions of CSIProxyService more recent than vN) along with an old version of csi-proxy.exe that does support vN.
+
+With that in mind, each subsection below details how to deprecate, then remove:
+
+#### A field in an API object
+
+Mark it as deprecated in the `proto` file, e.g.:
+```proto
+message ComputeDoubleRequest{
+    int32 input32 = 1 [deprecated=true];
+}
+```
+then regenerate the protobuf code.
+
+For removal, simply remove it in the first API version that doesn't support that field any more.
+
+#### An API procedure
+
+Similarly, mark the procedure as deprecated in the protobuf definition, e.g.:
+```proto
+service Dummy {
+    // ComputeDouble computes the double of the input. Real smart stuff!
+    rpc ComputeDouble(ComputeDoubleRequest) returns (ComputeDoubleResponse) {
+        option deprecated = true;
+    }
+}
+```
+then regenerate the protobuf code, and remove it from the first API version that doesn't support that procedure any more.
+
+#### An API version
+
+Again, mark the version as deprecated in its protobuf definition, e.g.:
+```proto
+// Deprecated: Do not use.
+// v1alpha1 is no longer maintained, and will be removed soon.
+package v1alpha1;
+
+service Dummy {
+    ...
+}
+```
+then regenerate the protobuf code, and run `csi-proxy-gen`: it will also mark the version's server (``) and client (``) packages as deprecated.
+
+For removal, remove the whole `api/<api_group_name>/<version>` directory, and run `csi-proxy-gen`, it will remove all references to the removed version.
+
+#### An API group
+
+Deprecate and remove all its versions as explained in the previous version; then remove the entire `api/<api_group_name>` directory, and run `csi-proxy-gen`, it will remove all references to the removed API group.
+
+## Detailed breakdown of generated files
+
+This section details how `csi-proxy-gen` works, and what files it generates; `csi-proxy-gen` is built on top of [gengo](https://github.com/kubernetes/gengo), and re-uses part of [k8s' code-generator](https://github.com/kubernetes/code-generator), notably to generate conversion functions.
```

Comment:
Do you see any risk of depending on these tools that were mainly designed for K8s APIs?  If changes are made in those tools that break our generation, will we need to fork them?

Reply:
That is indeed a very valid concern. The generator will mainly be based on gengo, whose purpose is well defined, and quite separate from k8s, so I think we're in the clear there.

As to code-generator, the only part that we're interested in is from https://github.com/kubernetes/code-generator/blob/master/cmd/conversion-gen/generators/conversion.go, namely comparing 2 structs, finding the fields with the same names and types in there, and generating efficient conversion code from one to the other. The plan is to make a PR on code-generator to extract the relevant pieces of code from conversion.go into a new package, and with extensive unit tests, and then use only that package in our own converter. Since that new package would have a well defined purpose, it should hopefully stay stable enough that we can rely on it without too much hassle going forward.